File: entry.py
#^
#^  HEAD
#^

#> HEAD -> MODULES
from dataclasses import dataclass
from pathlib import Path
import aiofiles
from typing import cast, Any


#^
#^  INPUT
#^

#> INPUT -> FILE
@dataclass(kw_only = True, frozen = True)
class File:
    name: Path
    async def read(self) -> str:
        async with aiofiles.open(self.name) as file: return await file.read()
    async def write(self, extension: str, content: str | bytes) -> None:
        async with aiofiles.open(self.name.with_suffix(extension), mode = "wb" if isinstance(content, bytes) else "w") as file: 
            await file.write(cast(Any, content))

#> INPUT -> FLAG
@dataclass(kw_only = True, frozen = True)
class Flag:
    value: str

#> INPUT -> ALIAS
@dataclass(kw_only = True, frozen = True)
class Alias:
    letters: str

#> INPUT -> TARGET
@dataclass(kw_only = True, frozen = True)
class Target:
    name: str

File: entry.rs
//^
//^ HEAD
//^

//> HEAD -> PRELUDE
use crate::prelude::{
    FilePath, 
    readFile, 
    writeFile,
    EnumString,
    AsRefStr
};

//> HEAD -> LOCAL
use super::issues::Issue;


//^
//^ ARGUMENT
//^

//> ARGUMENT -> ENUM
pub enum Argument {
    File(File),
    Flag(Flag),
    Alias(Alias),
    Target(Target)
}


//^
//^ INPUT
//^

//> INPUT -> FILE
#[derive(Clone)]
pub struct File (pub FilePath); impl File {
    pub fn read(&self) -> Result<String, Issue> {readFile(&self.0).ok().ok_or_else(|| Issue::FileNotFound(self.0.to_str().unwrap().to_string()))}
    pub fn write(&self, extension: &str, content: String) -> () {
        let mut path = self.0.clone();
        path.set_extension(extension);
        writeFile(path, content).unwrap();
    }
}

//> INPUT -> FLAG
#[derive(EnumString, Clone, Copy, AsRefStr)]
#[strum(serialize_all = "lowercase")]
pub enum Flag {
    Quiet,
    Verbose
}

//> INPUT -> ALIAS
#[derive(Clone, Debug)]
pub struct Alias(pub Vec<char>); impl Alias {pub fn expand(self) -> Result<Vec<Flag>, Issue> {
    let mut converted = Vec::new();
    for letter in &self.0 {converted.push(match letter {
        'q' => Flag::Quiet,
        'v' => Flag::Verbose,
        other => return Err(Issue::UnknownAliasCharacter {
            alias: self, 
            at: converted.len()
        })
    })}
    return Ok(converted);
}}

//> INPUT -> TARGET
#[derive(EnumString, Clone, AsRefStr, Copy)]
#[strum(serialize_all = "lowercase")]
pub enum Target {
    Help,
    Version,
    Tokens,
    Check,
    Ast,
    Latex
}